src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs

```diff
@@ -10,36 +10,47 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal static class BestTypeInferrer
     {
-        public static bool? GetIsNullable(ArrayBuilder<TypeSymbolWithAnnotations> types)
+        public static NullableAnnotation GetNullableAnnotation(ArrayBuilder<TypeSymbolWithAnnotations> types)
         {
-            bool? isNullable = false;
+            NullableAnnotation result = NullableAnnotation.NotNullableBasedOnAnalysis;
             foreach (var type in types)
             {
                 if (type.IsNull)
                 {
                     // https://github.com/dotnet/roslyn/issues/27961 Should ignore untyped
                     // expressions such as unbound lambdas and typeless tuples.
-                    isNullable = true;
+                    result = NullableAnnotation.NullableBasedOnAnalysis;
                     continue;
                 }
-                if (!type.IsReferenceType)
+
+                if (!type.IsReferenceType && !type.TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter())
                 {
-                    return null;
+                    return NullableAnnotation.Unknown;
                 }
-                switch (type.IsNullable)
+
+                NullableAnnotation nullableAnnotation = type.GetValueNullableAnnotation();
+
+                if (nullableAnnotation == NullableAnnotation.Unknown)
                 {
-                    case null:
-                        if (isNullable == false)
-                        {
-                            isNullable = null;
-                        }
-                        break;
-                    case true:
-                        isNullable = true;
-                        break;
+                    if (result.IsAnyNotNullable())
+                    {
+                        result = NullableAnnotation.Unknown;
+                    }
+                }
+                else if (nullableAnnotation.IsAnyNullable())
+                {
+                    if (!result.IsAnyNullable())
+                    {
+                        result = nullableAnnotation;
+                    }
+                    else if (result != nullableAnnotation)
+                    {
+                        result = NullableAnnotation.Nullable;
```

Comment:
>result = NullableAnnotation.Nullable; [](start = 24, length = 37)

A couple of comments here:
1) What is the purpose of setting `result` here? Couldn't we just leave it whatever nullable value it already has?

2) Not sure if it's observable, but if we need to set `result`, then it feels like we should use `NotNullableBasedOnAnalysis` here, as I think this method is only used for flow analysis.
Or is the idea to reserve `NotNullableBasedOnAnalysis` for unconstrained type parameters (where the nullability in analysis differs from that in binding)? If that's the case, I would add a comment on `NotNullableBasedOnAnalysis` enum value or rename it.

Reply:
> Why would we change nullable result to not-nullable? 

Sorry, I had mixed up Nullable and NotNullable in copy/paste. The question should be why not use `NullableBasedOnAnalysis`?

---
In reply to: [231258833](https://github.com/dotnet/roslyn/pull/30913#discussion_r231258833) [](ancestors = 231258833,231247378)